pkg/kube/sourcerepositories.go

```diff
@@ -39,13 +40,40 @@ func GetRepositoryGitURL(s *v1.SourceRepository) (string, error) {
 	return spec.HTTPCloneURL, nil
 }
 
-// FindSourceRepository returns a SourceRepository for the given namespace, owner and name
-// or returns an error if it cannot be found
-func FindSourceRepository(jxClient versioned.Interface, ns string, owner string, name string) (*v1.SourceRepository, error) {
+// FindSourceRepository returns a SourceRepository for the given namespace, owner, repo name, and (optional) provider name.
+// If no SourceRepository is found, return an error if doNotFailIfNotFound is false, otherwise just return a nil repo.
+func FindSourceRepository(jxClient versioned.Interface, ns string, owner string, name string, providerName string, doNotFailIfNotFound bool) (*v1.SourceRepository, error) {
```

Comment:
I would think we had a struct for repo provider + ower + name, but that is maybe not the case. There is `GitRepository` but that's too broad. Maybe we should add such a struct or we split `GitRepositor` in some way.

why the `doNotFailIfNotFound`? What do you try to achieve? I think it makes for a bad API. 
Why can the method not behave like:

* return repo + nil in case the source repository could be found
* return nil + nil if it could not be found (no error)
* return nil + error if any other error than apierrors.IsNotFound is encountered 

Are you trying to be be somewhat backwards compatible? Maybe in that case adding a new method makes more sense? Rename `FindSourceRepository` to `FindSourceRepositoryFailNotFound` then add a new implementation of FindSourceRepository which behaves as discussed above and use it. Eventually, we can then try to switch FindSourceRepositoryFailNotFound over to FindSourceRepository.

This is a general issue which I ran a lot into as well. These type of Find and Get operations we have written, fail to distinguish between not found and any other error. This makes it super tricky  use this methods.

Reply:
Actually, there's just one place that called it expecting an error if not found, so I'll fix that.